tensorflow/core/util/sycl_util.h

```diff
@@ -0,0 +1,98 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_USE_SYCL
+#error This file contains SYCL specific code and should only be included when \
+building with SYCL support.
+#endif
+
+#ifndef TENSORFLOW_CORE_UTIL_SYCL_UTIL_H_
+#define TENSORFLOW_CORE_UTIL_SYCL_UTIL_H_
+
+namespace tensorflow {
+
+// Need an atomic add for the MaxPoolGrad and MaxPool3DGrad kernels.
+//
+// For the device, this needs a pointer to global memory which isn't understood
+// by the host. The host should never be calling this method, but we provide
+// the header so that the host compiler can compile the functor.
+#ifdef __SYCL_DEVICE_ONLY__
+template <typename T>
+void SyclAtomicAdd(__attribute__((address_space(1))) T* address,
+                   const T increment);
+#else
+template <typename T>
+void SyclAtomicAdd(T* address, const T increment);
+#endif  // __SYCL_DEVICE_ONLY__
+
+#ifdef __SYCL_DEVICE_ONLY__
+// Use the OpenCL atomic uint operations to provide a floating point atomic add.
+// For the device we use the atomic compare-exchange builtin to keep trying to
+// add to the memory in a thread safe way. The union is needed as these
+// builtins are not availble for floating point types, only integer types, so
+// we do the addition on the float and the memory update on the uint.
+//
+// TODO(jwlawson): Remove once we have different type accessors for SYCL buffers
+// Providing a way to cast the types of buffers or accessors has been proposed
+// as a SYCL extension, so once this is available we can use an atomic
+// accessor and remove this.
```

Comment:
Access to the underlying buffers is requested as follow:
  auto input_access = input_buffer.template get_access<cl::sycl::access::mode::read>(cgh);
  auto output_access = output_buffer.template get_access<cl::sycl::access::mode::write>(cgh);
Can you explain why this can't be modifed into:
  auto input_access = input_buffer.template get_access<cl::sycl::access::mode::atomic>(cgh);
  auto output_access = output_buffer.template get_access<cl::sycl::access::mode::atomic>(cgh);